blocks/main_loop.py

```diff
@@ -109,13 +117,17 @@ def run(self):
             except TrainingFinish:
                 self.log.current_row.training_finished = True
             except Exception as e:
-                logger.error(traceback.format_exc(e))
-                logger.info(
-                    "An error occurred during the training.\n"
-                    "Attempting to run extensions before exiting...")
-                # TODO: change the serialization destination here
+                self.log.current_row.got_exception = traceback.format_exc(e)
+                logger.info(error_message)
+                try:
+                    self._run_extensions('on_error')
+                except Exception as inner_e:
+                    logger.error(traceback.format_exc(inner_e))
+                    logger.info(error_in_error_handling_message)
+                reraise_as(e)
```

Comment:
Strange... Maybe because of the second, nested `except` statement? Either way, leave it then